### DISCUSSION TITLE
API v2 authorization problem

### DISCUSSION THREAD

**Discussion by ForMiSoft** [tidal-music/discussions#84] [General]
API doesn't work even in the Tidal's web UI (please see the included screenshot bellow).
In my app I'm able to get Access Token from https://auth.tidal.com/v1/oauth2/token, but that's all we can do, everything else isn't working because there's an authentication/authorization problem on the Tidal's side. 

![API error](https://github.com/user-attachments/assets/faf07302-0397-47c8-9ab7-f305074001c6)

**Comment by aliaksei-taliuk** [comment]
Hey @NendosColl, thank you for reaching out.

We are not aware of any major issues with our V2 API. Our nightly regression tests pass for all currently available endpoints.

WRT to the specific issue you are facing, the error response you are seeing is most likely correct. I.e. the access token issued to the API Explorer has probably expired.

The API Explorer doesn't refresh access tokens on your behalf, but it currently requires you to re-authorize access to your resources when the access token expires. To do this, follow these steps:
1. Press "Exit API Explorer Mode" at the top-right corner
2. Press "Authorize" button
3. Press "Logout" under the PKCE section
4. Authorize again using PKCE

We acknowledge that this process is a bit tedious, and can probably be streamlined a bit, something we will look into down the road.

**Reply by ForMiSoft** [reply]
> Hey @NendosColl, thank you for reaching out.
> 
> We are not aware of any major issues with our V2 API. Our nightly regression tests pass for all currently available endpoints.
> 
> WRT to the specific issue you are facing, the error response you are seeing is most likely correct. I.e. the access token issued to the API Explorer has probably expired.
> 
> The API Explorer doesn't refresh access tokens on your behalf, but it currently requires you to re-authorize access to your resources when the access token expires. To do this, follow these steps:
> 
> 1. Press "Exit API Explorer Mode" at the top-right corner
> 2. Press "Authorize" button
> 3. Press "Logout" under the PKCE section
> 4. Authorize again using PKCE
> 
> We acknowledge that this process is a bit tedious, and can probably be streamlined a bit, something we will look into down the road.

Hi, thank you for your response. But my problem still persists.

![authorize](https://github.com/user-attachments/assets/35346ed3-492e-491a-ab48-b4784c1fbe42)

When I try to Authorize, there are two "Authorize" buttons (???), the right one does nothing and the left one responds with an error.

![error](https://github.com/user-attachments/assets/b172e22d-80a7-47c8-b340-919369cdeec0)

Any other ideas?

**Reply by aliaksei-taliuk** [reply]
Hi again @NendosColl ! Could you please:
1. Try to sign in into https://developer.tidal.com/
2. Go to Dashboard
3. Choose the application you're using
4. Go to App Settings
5. Ensure your app is allowed to claim these scopes (see Scopes section down below). Add 'user.read', 'entitlements.read', 'playlists.read' if not.

**Comment by ForMiSoft** [comment]
I really appreciate your help, but that wasn't the culprit.

![dashboard1](https://github.com/user-attachments/assets/5d7f689e-f434-46a6-a324-6494810dc931)

![dashboard2](https://github.com/user-attachments/assets/122b676b-09c3-4136-9552-ce2e048f6e46)

**Reply by ForMiSoft** [reply]
Wow, I tried it again and this time it worked. I don't know if you did something but thank you very much for your help, I almost gave up.

![success](https://github.com/user-attachments/assets/9241d6d7-6865-4761-a90f-163eb8b4cbdb)

**Comment by stefanmoro** [comment]
Glad to hear you got it working. Just another comment, I can see you have no redirect URIs configured for your app. Redirect URI is mandatory in order for you to be able to use authorization code flow in your client. Please read through the [authorization documentation](https://developer.tidal.com/documentation/authorization/authorization-overview) to learn more.

**Reply by ForMiSoft** [reply]
My joy was a little premature. Now it works in the Web UI, but my app is still getting errors when I try to obtain my playlists.

First I try to get an Access Token.

![get_access_token](https://github.com/user-attachments/assets/3961eee0-fdf2-4062-bda6-b7bd5ff2e754)

It works without a problem.  Is it ok that the returned scope is empty?

![access_token](https://github.com/user-attachments/assets/3661b75c-4d2b-493a-b2ac-c38be4fa7350)

Then I try to use that Access Token to get my playlists.

![get_playlists](https://github.com/user-attachments/assets/a86d10a7-92f0-4530-8bd3-8a33c87488e7)

It responds with an error 401 - Unauthorized

![error2](https://github.com/user-attachments/assets/85c96e6c-b4a1-4ea9-9a39-95a176518b95)

Please let me know what I'm doing wrong. Should I try to use a refresh token? If yes, how can I get that refresh token from my Windows desktop app?

**Reply by ForMiSoft** [reply]
> I can see you have no redirect URIs configured for your app. Redirect URI is mandatory in order for you to be able to use authorization code flow in your client.

My client is a Windows desktop app. I think that a redirect URI isn't very useful for an app that doesn't run on a web server.

**Comment by ForMiSoft** [comment]
I have a strong suspicion that there is no option in the API v2 to use its authorization flow from a Windows desktop app.

![sad-face-emotional](https://github.com/user-attachments/assets/989e6271-a1b3-47c6-a480-30a0401610c3)

**Reply by aliaksei-taliuk** [reply]
Hi @NendosColl . 

Please take a look here: https://github.com/orgs/tidal-music/discussions/6#discussioncomment-10392749

As it was mentioned there we're currently discussing the possibility to open up some of Playlist API endpoints (Public Playlists  only) for the Client Credentials flow. Private playlists or '/me' information will always work only under PKCE.

Best regards, 
Aliaksei
